src/PSR7/Validators/BodyValidator/MultipartValidation.php

```diff
@@ -0,0 +1,187 @@
+                            $partName,
+                            $partContentType,
+                            $addr
+                        );
+                    }
+                } else {
+                    // loose comparison (ie "image/*")
+                    $encodingContentType = str_replace('*', '.*', $encodingContentType);
+                    if (! preg_match('#' . $encodingContentType . '#', $partContentType)) {
+                        throw InvalidBody::becauseBodyDoesNotMatchSchemaMultipart(
+                            $partName,
+                            $partContentType,
+                            $addr
+                        );
+                    }
+                }
+
+                // 3.2. parts headers
+                foreach ($encoding->headers as $headerName => $headerSpec) {
+                    /** @var Header $headerSpec */
+                    $headerSchema = $headerSpec->schema;
+                    $headerValue  = $part->getHeader($headerName);
+
+                    if ($headerValue === null) {
+                        throw InvalidHeaders::becauseOfMissingRequiredHeaderMupripart($partName, $headerName, $addr);
+                    }
+
+                    $validator = new SchemaValidator($this->detectValidationStrategy($message));
+                    try {
+                        $validator->validate($headerValue, $headerSchema);
+                    } catch (SchemaMismatch $e) {
+                        throw InvalidHeaders::becauseValueDoesNotMatchSchemaMultipart($partName, $headerName, $headerValue, $addr, $e);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Prepare a Multipart message body (a set of normal parts) for validation against a schema
+     *
+     * @see https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html
+     * @see https://swagger.io/docs/specification/describing-request-body/multipart-requests/
+     *
+     * @return mixed[]
+     *
+     * @throws InvalidBody
+     * @throws TypeMismatch
+     */
+    private function parseMultipartData(OperationAddress $addr, StreamedPart $document) : array
+    {
+        $multipartData = []; // a buffer to fill up with message parts
+        foreach ($document->getParts() as $i => $part) {
+            $partContentType = $part->getHeader('Content-Type');
+
+            if (preg_match('#^application/.*json$#', $partContentType)) {
+                $partBody = json_decode($part->getBody(), true);
+                if (json_last_error() !== JSON_ERROR_NONE) {
+                    throw InvalidBody::becauseBodyIsNotValidJson(json_last_error_msg(), $addr);
+                }
+            } else {
+                $partBody = $part->getBody();
+            }
+
+            // if name is not set, it should be validated with "additionalProperties" keyword
+            $multipartData[$part->getName() ?? '____' . $i] = $partBody;
```

Comment:
how robust it is? what if i had part with a name `____`? (not sure whether it's possible, just imagine)

Reply:
I am not sure if a part can be unnamed at all. If it has no name set then we need to make one which must be unique... This code gives a name or `____N`.
This is a dirty solution, I agree.

Reply:
Since these are should be validated against additional properties than we can have separate array for them, having named and additional parts stored independently. this will make things more complex, yeah, but will be more correct and precise controller on the validation step